runtime/src/main/java/org/corfudb/runtime/collections/StreamNotificationTask.java

```diff
@@ -0,0 +1,98 @@
+ * to client via the registered call backs. This task is executed
+ * by the thread pool continuously until an error occurs which would
+ * stop the subscription.
+ * <p>
+ * Created by WenbinZhu on 11/9/20.
+ */
+@Slf4j
+class StreamNotificationTask implements Runnable {
+
+    // Number of transaction updates to send notification in each run.
+    private static final int NOTIFICATION_BATCH_SIZE = 10;
+
+    // Total amount of time to wait for retrieving the data changes from buffer if it is empty.
+    private static final Duration QUEUE_EMPTY_BLOCK_TIME_MS = Duration.ofMillis(1_000);
+
+    // A warning will be raised if client call back takes longer than this time threshold.
+    private static final Duration SLOW_NOTIFICATION_TIME_MS = Duration.ofMillis(1_500);
+
+    // The streaming manager that is in charge of listener subscriptions.
+    private final StreamingManager streamingManager;
+
+    // The subscription context associated with this task.
+    private final StreamSubscription subscription;
+
+    // The Thread pool for executing client notification tasks.
+    private final ExecutorService notificationExecutor;
+
+    StreamNotificationTask(StreamingManager streamingManager,
+                           StreamSubscription subscription,
+                           ExecutorService notificationExecutor) {
+        this.streamingManager = streamingManager;
+        this.subscription = subscription;
+        this.notificationExecutor = notificationExecutor;
+    }
+
+    @Override
+    public void run() {
+        try {
+            sendNotifications();
+        } catch (Throwable throwable) {
+            StreamListener listener = subscription.getListener();
+            log.error("Encountered exception {} during client notification callback, " +
+                    "listener: {}, namespace: {}", throwable, listener, subscription.getNamespace());
+            streamingManager.unsubscribe(listener, false);
+            listener.onError(throwable);
+        }
+    }
+
+    /**
+     * Retrieve the first data change from the buffer and send notification
+     * to the client via the pre-registered call back.
+     */
+    private void sendNotifications() throws Exception {
+        // Total amount of time for the batch waiting for buffer being not empty.
+        long remainingBlockTime = QUEUE_EMPTY_BLOCK_TIME_MS.toNanos();
+        StreamListener listener = subscription.getListener();
+
+        for (int iter = 0; iter < NOTIFICATION_BATCH_SIZE; iter++) {
+            // If listener already unsubscribed, do not process or schedule again.
+            if (subscription.isStopped()) {
+                return;
+            }
+
+            long startTime = System.nanoTime();
+            CorfuStreamEntries nextUpdate = subscription.dequeueStreamEntry(remainingBlockTime);
+
+            // No new updates after max waiting time elapses, break and re-schedule.
+            if (nextUpdate == null) {
+                break;
```

Comment:
Can you also add 

```java 

if (iter == 0 && nextUpdate == null) {

 notificationExecutor.schedule(this, IDLE_WAIT_TIME_MS, TimeUnit.MILLISECONDS);
return;
     }
```

This will prevent unnecessary waits on the consumer side when the stream is always empty. This improves the CPU usage for empty streams, but the ideal solution is to only submit the task if the producer enqueued entries. Going from 50ms polling to seconds should be ok for our scale, I think.

Reply:
Actually I think we do not need to wait in this case. If we hit this case, that means we already waited for some time, so cpu won't busy loop.

Reply:
I think it makes sense for the producer to wait one the enqueue to happen (i.e., on offer) because if the enqueue cannot happen right away it means means that the queue is full and the consumer is backed up and still processing. But when the consumer waits on poll, it never knows if the producer will ever enqueue or not, so it will keep waiting on the worker thread over and over again even if the stream is always empty.

 I feel like it makes to have `dequeueStreamEntry` return right away without waiting and then submitting the task with a delay in `notificationExecutor.schedule(this, delay)`